src/mbgl/style/expression/distance.cpp

```diff
@@ -137,25 +190,87 @@ double pointsToPointsDistance(const mapbox::geometry::multi_point<double>& point
     return dist;
 }
 
-std::pair<mbgl::optional<IndexRange>, mbgl::optional<IndexRange>> splitRange(const IndexRange& range, bool isLine) {
-    auto size = getRangeSize(range);
-    if (isLine) {
-        if (size == 2) {
-            return std::make_pair(range, nullopt);
+double pointToPolygonDistance(const mapbox::geometry::point<double>& point,
+                              const mapbox::geometry::polygon<double>& polygon,
+                              mapbox::cheap_ruler::CheapRuler& ruler) {
+    if (pointWithinPolygon(point, polygon, true /*trueOnBoundary*/)) return 0.0;
+    double dist = InfiniteDistance;
+    for (const auto& ring : polygon) {
+        const auto nearestPoint = std::get<0>(ruler.pointOnLine(ring, point));
+        dist = std::min(dist, ruler.distance(point, nearestPoint));
+        if (dist == 0.0) return dist;
+    }
+    return dist;
+}
+
+double lineToPolygonDistance(const mapbox::geometry::line_string<double>& line,
+                             const IndexRange& range,
+                             const mapbox::geometry::polygon<double>& polygon,
+                             mapbox::cheap_ruler::CheapRuler& ruler) {
+    if (!isRangeSafe(range, line.size())) {
+        mbgl::Log::Error(mbgl::Event::General, "index is out of range");
+        return InvalidDistance;
+    }
+    for (std::size_t i = range.first; i <= range.second; ++i) {
+        if (pointWithinPolygon(line[i], polygon, true /*trueOnBoundary*/)) return 0.0;
+    }
+
+    double dist = InfiniteDistance;
+    for (std::size_t i = range.first; i < range.second; ++i) {
+        const auto& p1 = line[i];
+        const auto& p2 = line[i + 1];
+        for (const auto& ring : polygon) {
+            for (std::size_t j = 0; j < ring.size() - 2; ++j) {
```

Comment:
should you use `ring.size() - 1` if the ring is just a container of points? You should also check the last segment between `ring[0]` and `ring[ring.size()] - 1`